### DISCUSSION TITLE
explainer selection

### DISCUSSION THREAD

**Discussion by amylee-lixinyi** [shap/shap#3198] [Q&A]
I'm relatively new to SHAP package, documents reading and googling failed to answering me which SHAP explainer to use when try to interprete multiple kinds of ML models, for example tree based models, SVR and ANN. Based on this [post](https://stackoverflow.com/questions/73329209/difference-between-shap-treeexplainer-and-shap-explainer-bar-charts) and this [post](https://stackoverflow.com/questions/74251331/what-is-difference-between-explainer-and-kernelexplainer-in-shap) `shap.Explainer` is a superset of other explainers like`shap.TreeExplainer`, if so, when should subset explainer be use instead of `shap.Explainer`, whether it means stick to `shap.Explainer` is the safer choice anytime?

**Comment by thatlittleboy** [comment]
Hi Amy, `shap.Explainer` is just a general wrapper over the other subclasses like `shap.TreeExplainer`. I'ld say if you know what model type you are using, then go with the more specific one rather than relying on `shap.Explainer`'s auto-inference. I.e., TreeExplainer in this case.

**Reply by amylee-lixinyi** [reply]
Hi thanks for your swift answer. If so, my question will be which subclasses explainer should I choose, say for SVR, ANN and RF?  I guess `shap.TreeExplainer` might be the one fit for tree based models, but `shap.KernelExplainer` can deal with tree based models as a model agnostic explainer. Also, if I want to compare the feature importance between different ML models, should I stick to a single explainer to ensure the comparable of sharply values between models?

**Reply by abarbosa94** [reply]
Hi @amylee-lixinyi,

> but shap.KernelExplainer can deal with tree-based models as a model-agnostic explainer

This is correct, but bear in mind that KernelExplainer is much more expensive to compute than TreeExplainer.

> Should I stick to a single explainer to ensure the comparison of sharp values between models?

This is a good question. Shapley Values are unique solutions to feature importance computation and are achievable through ExactExplainer API. I can be wrong here, but AFAIK, **all the other APIs** are approximate solutions to the Shapley value equation. Therefore, as they are trying to achieve the same thing (computing the Shapley values, which are unique), I believe that they are comparable.

If you want more robustness though, I'd prefer to use a single Explainer to everything

**Reply by amylee-lixinyi** [reply]
Hi @abarbosa94 thanks for your insight into this.
If different explainers are trying to approximate the same Shapley value, how to choose between them, say for model agnostic explainer, kernel and permutation explainers are fit for the job.

**Reply by abarbosa94** [reply]
Hi @amylee-lixinyi, 

That's an interesting question. I'd say it is a balance between the precision you want to achieve and the computation resources you have in mind.

- Model Agnostic Explainer and KernelExplainer are the same thing. They are the most general and expensive ones to compute, requiring $O(2^M)$
- TreeExplainers only work for Tree related models, and they are blazingly fast
- Partition Explainer is faster than KernelExplainer, but it assumes that there are correlations between your features, which makes sense to some applications such as Text and Images. When we have a balanced tree as the hierarchy of features, it is $O(M^2)$, M being the number of features
- Permutation Explainer is another model agnostic model, but I'm still trying to understand its nuances as raised a question here #3220 

If you want to learn more about it, the benchmark section in the documentation provides some good intuition: https://shap.readthedocs.io/en/stable/example_notebooks/benchmarks/tabular/Benchmark%20XGBoost%20explanations.html
